File: agents/render/layout.py
"""Mise en page produit — adapte une œuvre de ratio « musée » aux tailles POD.

Problème : Gelato/Prodigi ne fabriquent pas de cadre au ratio exact de l'œuvre —
ils travaillent sur un CATALOGUE de tailles fixes. Une reproduction patrimoniale
ne doit JAMAIS être rognée (crop) ni déformée (stretch). La seule voie fidèle :
contenir l'œuvre à son ratio natif et combler le reste par une BORDURE / un
PASSE-PARTOUT — soit intégré au fichier (Gelato), soit physique (mat Prodigi).

Ce module choisit la taille catalogue dont le ratio est le plus proche (bordure
minimale, équilibrée) et calcule les marges exactes. Voir docs/decision-encadrement-tailles.md.

Sources POD : cf. docs/benchmark-pod-fournisseurs.md §6 + todo §3.2/3.3 (à confirmer
par mail pour Gelato : libellé fit/fill, bornes custom).
"""

from __future__ import annotations

import math
from dataclasses import dataclass
from typing import List, Optional, Tuple


@dataclass(frozen=True)
class TailleCatalogue:
    label: str
    court_cm: float        # petit côté
    long_cm: float         # grand côté
    fournisseurs: Tuple[str, ...]   # ("gelato", "prodigi")

    @property
    def ratio(self) -> float:       # toujours ≥ 1 (grand/petit)
        return self.long_cm / self.court_cm


# Tailles murales réelles, convergentes Gelato/Prodigi (cm). Ratio = grand/petit.
CATALOGUE: List[TailleCatalogue] = [
    TailleCatalogue("20×25", 20, 25, ("prodigi",)),               # 8×10"  · 1.25
    TailleCatalogue("28×36", 28, 35.6, ("prodigi",)),             # 11×14" · 1.27
    TailleCatalogue("30×40", 30, 40, ("gelato", "prodigi")),      # 12×16" · 1.33
    TailleCatalogue("A3 30×42", 29.7, 42, ("gelato",)),           # 1.414
    TailleCatalogue("40×50", 40, 50, ("gelato", "prodigi")),      # 16×20" · 1.25
    TailleCatalogue("50×70", 50, 70, ("gelato", "prodigi")),      # ~20×28" · 1.40
    TailleCatalogue("A2 42×59", 42, 59.4, ("gelato",)),           # 1.414
    TailleCatalogue("61×91", 61, 91.4, ("gelato", "prodigi")),    # 24×36" · 1.50
    TailleCatalogue("30×30", 30, 30, ("gelato", "prodigi")),      # carré 1.0
    TailleCatalogue("50×50", 50, 50, ("gelato", "prodigi")),      # carré 1.0
]


@dataclass
class PlanMiseEnPage:
    taille: str            # libellé catalogue
    fournisseur: str       # "gelato" | "prodigi"
    orientation: str       # "paysage" | "portrait" | "carré"
    ratio_oeuvre: float    # grand/petit de l'œuvre
    ratio_taille: float    # grand/petit de la taille catalogue
    marge_h_frac: float    # bordure gauche+droite, en fraction de la largeur d'œuvre
    marge_v_frac: float    # bordure haut+bas, en fraction de la hauteur d'œuvre
    bordure_pct: float     # part de la surface finale occupée par la bordure (0-100)
    sizing_prodigi: str    # comportement API recommandé
    strategie: str         # "bordure-fichier" (Gelato) | "passe-partout" (Prodigi)
    note: str

    def to_dict(self) -> dict:
        from dataclasses import asdict
        return asdict(self)


def _orientation(ratio_brut: float) -> str:
    if ratio_brut > 1.05:
        return "paysage"
    if ratio_brut < 0.95:
        return "portrait"
    return "carré"


def ouverture(aw: int, ah: int, ratio_cible: float, mat_min_frac: float = 0.08
              ) -> Tuple[int, int, int, int]:
    """Géométrie de l'ouverture (œuvre + passe-partout) au ``ratio_cible`` exact,
    contenant l'œuvre + au moins ``mat_min_frac`` du petit côté en marge par bord.

    ``ratio_cible`` est exprimé dans l'ORIENTATION de l'œuvre (largeur/hauteur).
    Retourne (ouverture_w, ouverture_h, marge_x_px, marge_y_px). Jamais de crop.
    """
    m = int(min(aw, ah) * mat_min_frac)
    # ouverture_w / ouverture_h = ratio_cible, avec marge ≥ m sur chaque axe.
    oh = max(ah + 2 * m, math.ceil((aw + 2 * m) / ratio_cible))
    ow = max(aw + 2 * m, round(ratio_cible * oh))
    oh = max(oh, round(ow / ratio_cible))  # cohérence après arrondis
    mwx = (ow - aw) // 2
    mwy = (oh - ah) // 2
    return ow, oh, mwx, mwy


def meilleure_taille(ratio_oeuvre: float, fournisseur: Optional[str] = None
                     ) -> TailleCatalogue:
    """Taille catalogue dont le ratio est le plus proche de l'œuvre (bordure minimale).

    ``ratio_oeuvre`` ≥ 1 (grand/petit). Filtre par fournisseur si fourni.
    """
    cands = [t for t in CATALOGUE if (fournisseur is None or fournisseur in t.fournisseurs)]
    # exclure le carré sauf si l'œuvre est ~carrée (sinon bordure énorme)
    if ratio_oeuvre > 1.12:
        cands = [t for t in cands if t.ratio > 1.05] or cands
    return min(cands, key=lambda t: abs(t.ratio - ratio_oeuvre))


def planifier(aw: int, ah: int, fournisseur: str = "gelato",
              taille: Optional[TailleCatalogue] = None,
              mat_min_frac: float = 0.08) -> PlanMiseEnPage:
    """Plan de mise en page d'une œuvre ``aw×ah`` pour un fournisseur.

    Choisit la taille (la plus proche si non imposée), calcule les marges au ratio
    cible et recommande le comportement de mise à l'échelle. Aucune perte d'image.
    """
    ratio_brut = aw / ah
    orient = _orientation(ratio_brut)
    ratio_oeuvre = max(aw, ah) / min(aw, ah)
    if taille is None:
        taille = meilleure_taille(ratio_oeuvre, fournisseur)

    # ratio cible dans l'orientation de l'œuvre
    ratio_cible = taille.ratio if orient != "portrait" else 1.0 / taille.ratio
    if orient == "carré":
        ratio_cible = 1.0
    ow, oh, mwx, mwy = ouverture(aw, ah, ratio_cible, mat_min_frac)
    marge_h_frac = round(2 * mwx / aw, 3)
    marge_v_frac = round(2 * mwy / ah, 3)
    bordure_pct = round(100 * (1 - (aw * ah) / (ow * oh)), 1)

    if fournisseur == "prodigi":
        strategie = "passe-partout"
        sizing = "fitPrintArea (jamais de crop) + mount activé"
        note = ("Prodigi : commander en fitPrintArea avec passe-partout (mat) — l'œuvre "
                "garde son ratio, le mat physique fournit la marge. Choisir la couleur "
                "de mat (snow white).")
    else:
        strategie = "bordure-fichier"
        sizing = "fillPrintArea (le fichier est déjà au ratio exact → aucun crop)"
        note = ("Gelato : composer le fichier print à la taille catalogue avec la bordure "
                "blanche intégrée (faux passe-partout). Le ratio du fichier = celui du "
                "produit → pas de crop fournisseur.")

    return PlanMiseEnPage(
        taille=taille.label, fournisseur=fournisseur, orientation=orient,
        ratio_oeuvre=round(ratio_oeuvre, 3), ratio_taille=round(taille.ratio, 3),
        marge_h_frac=marge_h_frac, marge_v_frac=marge_v_frac, bordure_pct=bordure_pct,
        sizing_prodigi=sizing, strategie=strategie, note=note,
    )


def plans_variants(aw: int, ah: int, fournisseur: str = "gelato") -> List[PlanMiseEnPage]:
    """Un plan par taille catalogue offerte par le fournisseur (variantes du listing)."""
    return [planifier(aw, ah, fournisseur, t)
            for t in CATALOGUE if fournisseur in t.fournisseurs]

File: web/lib/layout.ts
/**
 * Mise en page produit — port TypeScript fidèle de `agents/render/layout.py`.
 *
 * Une œuvre de ratio « musée » ne doit JAMAIS être rognée ni déformée : on
 * choisit la taille catalogue POD dont le ratio est le plus proche (bordure
 * minimale) et on comble par une bordure/passe-partout. Ce module permet au
 * cockpit et au constructeur de listing de calculer les variantes FIDÈLES
 * à offrir pour CHAQUE œuvre (selon son ratio), sans appeler Python.
 *
 * Géométrie ENTIÈRE identique à layout.py : taille catalogue, orientation et
 * ouverture/marges en px partagent le même arrondi banquier (`pyRound`). Les
 * valeurs d'AFFICHAGE arrondies (ratio, bordure %) peuvent différer d'un cran
 * sur un demi exact (`toFixed` arrondit away-from-zero, Python `round` vers le
 * pair) — cosmétique, sans incidence sur le choix de taille. Symétrie vérifiée
 * par tests/test_layout_ts_symmetry.py (exécute ce port via Node). Voir
 * docs/decision-encadrement-tailles.md.
 */

export type Fournisseur = "gelato" | "prodigi";

export type TailleCatalogue = {
  label: string;
  courtCm: number; // petit côté
  longCm: number; // grand côté
  fournisseurs: readonly Fournisseur[];
};

export function ratioTaille(t: TailleCatalogue): number {
  // toujours ≥ 1 (grand/petit)
  return t.longCm / t.courtCm;
}

// Tailles murales réelles, convergentes Gelato/Prodigi (cm) — copie de layout.CATALOGUE.
export const CATALOGUE: readonly TailleCatalogue[] = [
  { label: "20×25", courtCm: 20, longCm: 25, fournisseurs: ["prodigi"] }, // 8×10" · 1.25
  { label: "28×36", courtCm: 28, longCm: 35.6, fournisseurs: ["prodigi"] }, // 11×14" · 1.27
  { label: "30×40", courtCm: 30, longCm: 40, fournisseurs: ["gelato", "prodigi"] }, // 12×16" · 1.33
  { label: "A3 30×42", courtCm: 29.7, longCm: 42, fournisseurs: ["gelato"] }, // 1.414
  { label: "40×50", courtCm: 40, longCm: 50, fournisseurs: ["gelato", "prodigi"] }, // 16×20" · 1.25
  { label: "50×70", courtCm: 50, longCm: 70, fournisseurs: ["gelato", "prodigi"] }, // ~20×28" · 1.40
  { label: "A2 42×59", courtCm: 42, longCm: 59.4, fournisseurs: ["gelato"] }, // 1.414
  { label: "61×91", courtCm: 61, longCm: 91.4, fournisseurs: ["gelato", "prodigi"] }, // 24×36" · 1.50
  { label: "30×30", courtCm: 30, longCm: 30, fournisseurs: ["gelato", "prodigi"] }, // carré 1.0
  { label: "50×50", courtCm: 50, longCm: 50, fournisseurs: ["gelato", "prodigi"] }, // carré 1.0
];

export type Orientation = "paysage" | "portrait" | "carré";

export type PlanMiseEnPage = {
  taille: string; // libellé catalogue
  fournisseur: Fournisseur;
  orientation: Orientation;
  ratioOeuvre: number; // grand/petit de l'œuvre
  ratioTaille: number; // grand/petit de la taille catalogue
  margeHFrac: number; // bordure gauche+droite, fraction de la largeur d'œuvre
  margeVFrac: number; // bordure haut+bas, fraction de la hauteur d'œuvre
  bordurePct: number; // part de la surface finale occupée par la bordure (0-100)
  strategie: "bordure-fichier" | "passe-partout";
};

function orientationDe(ratioBrut: number): Orientation {
  if (ratioBrut > 1.05) return "paysage";
  if (ratioBrut < 0.95) return "portrait";
  return "carré";
}

/** round() de Python : arrondi au plus proche, .5 exact → vers le PAIR (Math.round arrondit vers le haut). */
function pyRound(x: number): number {
  const floor = Math.floor(x);
  if (x - floor === 0.5) return floor % 2 === 0 ? floor : floor + 1;
  return Math.round(x);
}

/**
 * Géométrie de l'ouverture (œuvre + passe-partout) au ratio cible exact,
 * contenant l'œuvre + au moins `matMinFrac` du petit côté en marge par bord.
 * Identique à layout.ouverture (jamais de crop), mêmes arrondis que Python.
 */
export function ouverture(
  aw: number,
  ah: number,
  ratioCible: number,
  matMinFrac = 0.08,
): { ow: number; oh: number; margeX: number; margeY: number } {
  const m = Math.floor(Math.min(aw, ah) * matMinFrac);
  let oh = Math.max(ah + 2 * m, Math.ceil((aw + 2 * m) / ratioCible));
  const ow = Math.max(aw + 2 * m, pyRound(ratioCible * oh));
  oh = Math.max(oh, pyRound(ow / ratioCible)); // cohérence après arrondis
  return { ow, oh, margeX: Math.floor((ow - aw) / 2), margeY: Math.floor((oh - ah) / 2) };
}

/** Taille catalogue dont le ratio est le plus proche de l'œuvre (bordure minimale). */
export function meilleureTaille(ratioOeuvre: number, fournisseur?: Fournisseur): TailleCatalogue {
  let cands = CATALOGUE.filter((t) => !fournisseur || t.fournisseurs.includes(fournisseur));
  // exclure le carré sauf si l'œuvre est ~carrée (sinon bordure énorme)
  if (ratioOeuvre > 1.12) {
    const nonCarre = cands.filter((t) => ratioTaille(t) > 1.05);
    if (nonCarre.length) cands = nonCarre;
  }
  return cands.reduce((best, t) =>
    Math.abs(ratioTaille(t) - ratioOeuvre) < Math.abs(ratioTaille(best) - ratioOeuvre) ? t : best,
  );
}

// Arrondi DÉCIMAL d'affichage. toFixed évite l'amplification d'erreur flottante
// de Math.round(x*10)/10 ; il diverge de round() banquier de Python uniquement
// sur un demi exact (away-from-zero vs vers-le-pair) → écart d'un cran possible
// sur ratio/bordure %. Cosmétique : taille/orientation/géométrie px n'en dépendent pas.
const round3 = (n: number) => Number(n.toFixed(3));
const round1 = (n: number) => Number(n.toFixed(1));

/** Plan de mise en page d'une œuvre aw×ah px pour un fournisseur (cf. layout.planifier). */
export function planifier(
  aw: number,
  ah: number,
  fournisseur: Fournisseur = "gelato",
  taille?: TailleCatalogue,
  matMinFrac = 0.08,
): PlanMiseEnPage {
  const ratioBrut = aw / ah;
  const orient = orientationDe(ratioBrut);
  const ratioOeuvre = Math.max(aw, ah) / Math.min(aw, ah);
  const t = taille ?? meilleureTaille(ratioOeuvre, fournisseur);

  let ratioCible = orient === "portrait" ? 1 / ratioTaille(t) : ratioTaille(t);
  if (orient === "carré") ratioCible = 1;
  const { ow, oh, margeX, margeY } = ouverture(aw, ah, ratioCible, matMinFrac);

  return {
    taille: t.label,
    fournisseur,
    orientation: orient,
    ratioOeuvre: round3(ratioOeuvre),
    ratioTaille: round3(ratioTaille(t)),
    margeHFrac: round3((2 * margeX) / aw),
    margeVFrac: round3((2 * margeY) / ah),
    bordurePct: round1(100 * (1 - (aw * ah) / (ow * oh))),
    strategie: fournisseur === "prodigi" ? "passe-partout" : "bordure-fichier",
  };
}

/** Un plan par taille catalogue offerte par le fournisseur (variantes du listing). */
export function plansVariants(aw: number, ah: number, fournisseur: Fournisseur = "gelato"): PlanMiseEnPage[] {
  return CATALOGUE.filter((t) => t.fournisseurs.includes(fournisseur)).map((t) =>
    planifier(aw, ah, fournisseur, t),
  );
}
